src/app/modules/auth/auth.service.ts

```diff
@@ -339,3 +348,50 @@ export const getFormIfPublic = (
     FormService.isFormPublic(form).map(() => form),
   )
 }
+
+/**
+ * Retrieves the user of the given API key
+ *
+ * @returns ok(IUserSchema) if the API key matches the hashed API key in the DB
+ * @returns err(DatabaseError) if database errors occurs whilst retrieving user
+ * @returns err(MissingUserError) if user does not exist in the database
+ */
+export const getUserByApiKey = (
+  userId: string,
+  token: string,
+): ResultAsync<IUserSchema, Error> => {
+  return findUserById(userId).andThen((user) => {
+    if (!user.apiKeyHash) {
+      return errAsync(new MissingTokenError())
+    }
+    return compareHash(token, user.apiKeyHash ?? '').andThen((isHashMatch) => {
+      if (isHashMatch) return okAsync(user)
+      return errAsync(new InvalidTokenError())
+    })
+  })
+}
+
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+const getApiKeyHash = (apiKey: string): ResultAsync<string, HashingError> => {
+  return ResultAsync.fromPromise(
+    bcrypt.hash(apiKey, DEFAULT_SALT_ROUNDS),
+    (error) => {
+      logger.error({
+        message: 'bcrypt hash error',
+        meta: {
+          action: 'getApiKeyHash',
+        },
+        error,
+      })
+      return new HashingError()
+    },
+  ).map((hash) => `${hash}`)
+}
+
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+const generateApiKey = (user: IUserSchema): string => {
+  const key = crypto.randomBytes(32).toString('base64')
+  const apiEnv = config.publicApiConfig.apiEnv
+  const apiKeyVersion = config.publicApiConfig.apiKeyVersion
+  return `${apiEnv}_${apiKeyVersion}_${user._id}_${key}`
```

Comment:
During the huddle, brought up that using `user._id` without obfuscation is kinda the same as using `email` as an identifier. But on second thought, user id is slightly better in the case where attackers somehow have a list of api keys, then the user ids provide them with an easier vector for them to figure out which agency they want to attack haha